### ISSUE TITLE
DeploymentClusteredTest.shouldDeployProcessAndCreateInstances

### ISSUE THREAD

**Issue by Zelldon** [camunda-cloud/zeebe#7981]
**Summary**

> Try to answer the following as best as possible

- How often does the test fail? at least once
- Does it block your work? Blocked merging
- Do we suspect that it is a real failure? not sure

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
io.camunda.zeebe.it.clustering.DeploymentClusteredTest

[2021-10-13T09:22:45.475Z] [ERROR] io.camunda.zeebe.it.clustering.DeploymentClusteredTest.shouldDeployProcessAndCreateInstances  Time elapsed: 120.012 s  <<< ERROR!

[2021-10-13T09:22:45.475Z] org.junit.runners.model.TestTimedOutException: test timed out after 120 seconds

[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/jdk.internal.misc.Unsafe.park(Native Method)

[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:234)

[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2123)

[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.util.sched.future.CompletableActorFuture.get(CompletableActorFuture.java:133)

[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.util.sched.future.CompletableActorFuture.get(CompletableActorFuture.java:108)

[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.util.sched.FutureUtil.join(FutureUtil.java:21)

[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.util.sched.future.CompletableActorFuture.join(CompletableActorFuture.java:196)

[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.broker.Broker.lambda$close$0(Broker.java:170)

[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.broker.Broker$$Lambda$1358/0x00000007c0977c40.accept(Unknown Source)

[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.CompletableFuture.uniAcceptNow(CompletableFuture.java:753)

[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:731)

[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2108)
[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.broker.Broker.lambda$close$1(Broker.java:168)
[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.broker.Broker$$Lambda$1357/0x00000007c0977840.run(Unknown Source)
[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.util.LogUtil.doWithMDC(LogUtil.java:23)
[2021-10-13T09:22:45.475Z] 	at app//io.camunda.zeebe.broker.Broker.close(Broker.java:163)
[2021-10-13T09:22:45.475Z] 	at app/zeebe.qa.integration.tests@1.3.0-SNAPSHOT/io.camunda.zeebe.it.clustering.ClusteringRule.lambda$after$3(ClusteringRule.java:257)
[2021-10-13T09:22:45.475Z] 	at app/zeebe.qa.integration.tests@1.3.0-SNAPSHOT/io.camunda.zeebe.it.clustering.ClusteringRule$$Lambda$1582/0x00000007c0a67840.accept(Unknown Source)
[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1675)
[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
[2021-10-13T09:22:45.475Z] 	at java.base@11.0.10/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurrent.ForkJoinPool$WorkQueue.helpCC(ForkJoinPool.java:1115)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurrent.ForkJoinPool.externalHelpComplete(ForkJoinPool.java:1957)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurent.ForkJoinTask.tryExternalHelp(ForkJoinTask.java:378)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurrent.ForkJoinTask.externalAwaitDone(ForkJoinTask.java:323)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:412)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:736)
[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)

[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)

[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)

[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)

[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:661)

[2021-10-13T09:22:45.476Z] 	at app/zeebe.qa.integration.tests@1.3.0-SNAPSHOT/io.camunda.zeebe.it.clustering.ClusteringRule.after(ClusteringRule.java:254)

[2021-10-13T09:22:45.476Z] 	at app//org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:59)

[2021-10-13T09:22:45.476Z] 	at app//org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)

[2021-10-13T09:22:45.476Z] 	at app//org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)

[2021-10-13T09:22:45.476Z] 	at app//org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)

[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.util.concurrent.FutureTask.run(FutureTask.java:264)

[2021-10-13T09:22:45.476Z] 	at java.base@11.0.10/java.lang.Thread.run(Thread.java:834)

[2021-10-13T09:22:45.476Z] 
</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted

**Logs**

> If possible, provide more context here, e.g. standard output logs, link to build, etc.

<details><summary>Logs</summary>
<pre>
LOGS
</pre>
</details>

Failing https://ci.zeebe.camunda.cloud/blue/organizations/jenkins/camunda-cloud%2Fzeebe/detail/staging/1173/pipeline

**Comment by deepthidevaki** [comment]
Was able to reproduce this and observed the following:

TLDR: On Actor closing some jobs are discarded, as a result transitionToInactive waits for ever and ZeebePartition is not closed.


ZeebePartition::closeAsync() is executed. (Following log is added by me for debugging)
```
13:36:52.275 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-1] INFO  io.camunda.zeebe.broker.system - FINDME: Scheduling close after currentTransition is completed
```
currentTransition = transition to Follower.

New transition to leader is requested (triggered by raft role change)

```
13:36:52.278 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-1] INFO  io.camunda.zeebe.broker.system - Transition to LEADER on term 2 requested.
13:36:52.278 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-1] DEBUG io.camunda.zeebe.broker.system - Partition role transitioning from CANDIDATE to LEADER in term 2
13:36:52.278 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-1] INFO  io.camunda.zeebe.broker.system - Received cancel signal for transition to FOLLOWER on term 2
```

currentTransition will be cancelled.

```
13:36:52.283 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.camunda.zeebe.broker.system - Detected 'HEALTHY' components. The current health status of components: [ZeebePartition-1{status=HEALTHY}, raft-partition-partition-1{status=HEALTHY}, Broker-1-LogStream-1{status=HEALTHY}, Broker-1-StreamProcessor-1{status=HEALTHY}]
13:36:52.283 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] INFO  io.camunda.zeebe.broker.system - Cancelling transition to FOLLOWER on term 2
13:36:52.283 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] INFO  io.camunda.zeebe.broker.system - ZeebePartition-1 recovered, marking it as healthy
13:36:52.283 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.camunda.zeebe.broker.system - Detected 'HEALTHY' components. The current health status of components: [ZeebePartition-1{status=HEALTHY}, raft-partition-partition-1{status=HEALTHY}, Broker-1-LogStream-1{status=HEALTHY}, Broker-1-StreamProcessor-1{status=HEALTHY}]
```
currentTransition is cancelled => completed

```
13:36:52.283 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] INFO  io.camunda.zeebe.broker.system - FINDME: Current transition completed. calling super.closeAsync.
```
Since currentTransition is cancelled, we start the closing the actor.

But another transition is enqueued, and it starts the transition.
```
13:36:52.283 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] INFO  io.camunda.zeebe.broker.system - Prepare transition from FOLLOWER on term 2 to LEADER
13:36:52.283 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] INFO  io.camunda.zeebe.broker.system - Prepare transition from FOLLOWER on term 2 to LEADER - preparing ExporterDirector
```

But since we have already called actor.closeAsync(), it discards the enqueued jobs.
```
13:36:52.283 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.camunda.zeebe.util.sched.ActorTask - Discard job io.camunda.zeebe.util.sched.future.FutureContinuationRunnable QUEUED from fastLane of Actor Broker-1-ZeebePartition-1.
13:36:52.283 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.camunda.zeebe.util.sched.ActorTask - Discard job io.camunda.zeebe.util.sched.future.FutureContinuationRunnable QUEUED from fastLane of Actor Broker-1-ZeebePartition-1.
```
So tasks related to transition to LEADER is discarded, as a result that transition never gets compelted.


onActorClosing we transition to Inactive
```
13:36:52.283 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] DEBUG io.camunda.zeebe.broker.system - Closing ZeebePartition 1
13:36:52.283 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] INFO  io.camunda.zeebe.broker.system - Transition to INACTIVE on term 2 requested.
13:36:52.283 [Broker-1-ZeebePartition-1] [Broker-1-zb-actors-0] INFO  io.camunda.zeebe.broker.system - Received cancel signal for transition to LEADER on term 2
```
But transition to Inactive never gets executed because the previous transition to LEADER is never completed.
